include/kframework/infinite-gas.md

```diff
@@ -5,134 +5,55 @@ Here we use the construct `#gas` to represent positive infinity, while tracking
-endmodule
-
-module INFINITE-GAS-HASKELL [kore]
-    imports INFINITE-GAS-COMMON
-
-    rule #gas(_) <=Int _C => false [concrete(_C), simplification]
-    rule _C <=Int #gas(_) => true  [concrete(_C), simplification]
-
-    rule { #gas(_) #Equals #gas(_) } => #Top [simplification]
-endmodule
-
-module INFINITE-GAS-COMMON
+    imports BASIC-K
+    imports INT
+    imports BOOL
+    imports GAS
     imports EVM
 
- // Generic rules
+    syntax Gas ::= #gas(Int)
 
-    rule (notBool (A andBool B)) andBool A => (notBool B) andBool A                      [simplification]
-    rule #if B #then C +Int C1 #else C +Int C2 #fi => C +Int #if B #then C1 #else C2 #fi [simplification]
+    rule #gas(G) +Gas G'       => #gas(G +Int G')
+    rule #gas(G) -Gas G'       => #gas(G -Int G')
+    rule #gas(G) *Gas G'       => #gas(G *Int G')
+    rule #gas(G) /Gas G'       => #gas(G /Int G')  requires G' =/=Int 0
+    rule      G  +Gas #gas(G') => #gas(G +Int G')
+    rule      G  -Gas #gas(G') => #gas(G -Int G')
+    rule      G  *Gas #gas(G') => #gas(G *Int G')
+    rule      G  /Gas #gas(G') => #gas(G /Int G')  requires G' =/=Int 0
+    rule #gas(G) +Gas #gas(G') => #gas(G +Int G')
+    rule #gas(G) -Gas #gas(G') => #gas(G -Int G')
+    rule #gas(G) *Gas #gas(G') => #gas(G *Int G')
+    rule #gas(G) /Gas #gas(G') => #gas(G /Int G')  requires G' =/=Int 0
 
- // Symbolic Gas
+    rule  #gas(_)  <Gas _:Int   => false
+    rule  #gas(_)  <Gas #gas(_) => false
+    rule  _:Int    <Gas #gas(_) => true
 
-    syntax Int ::= #gas ( Int ) [function, total, no-evaluators, klabel(infGas), symbol, smtlib(infGas)]
- // ----------------------------------------------------------------------------------------------------
-    rule #gas(G) +Int G' => #gas(G +Int G') requires 0 <=Int G' orBool 0 -Int G' <Int #gas(G)  [simplification]
-    rule G +Int #gas(G') => #gas(G +Int G') requires 0 <=Int G  orBool 0 -Int G  <Int #gas(G') [simplification]
+    rule  #gas(_) <=Gas _:Int   => false
+    rule  #gas(_) <=Gas #gas(_) => true
+    rule  _:Int   <=Gas #gas(_) => true
 
-    rule #gas(G) -Int G'       => #gas(G -Int G') requires 0 <=Int G' andBool G' <Int #gas(G) [simplification]
-    rule #gas(G) -Int #gas(G') => #gas(G -Int G')                                             [simplification]
+    rule minGas(#gas(G), #gas(G')) => #gas(minInt(G, G'))
+    rule minGas(G:Int  , #gas(_))  => G
+    rule minGas(#gas(_), G':Int)   => G'
 
-    rule #gas(G) *Int G' => #gas(G *Int G') requires 0 <=Int G' [simplification]
-    rule G *Int #gas(G') => #gas(G *Int G') requires 0 <=Int G  [simplification]
+    rule gas2Int(#gas(G)) => G
 
-    rule #gas(G) /Int G' => #gas(G /Int G') requires 0  <Int G' andBool G' <Int #gas(G)  [simplification]
-    rule G /Int #gas(G') => 0               requires 0 <=Int G  andBool G  <Int #gas(G') [simplification]
+    rule _ <=Gas #gas(_)        => true  [simplification]
+    rule         #gas(_) <Gas _ => false [simplification]
```

Comment:
@ehildenb Would the rules on 49 and 50 speed things up in the event that `_` is a function? Previously I was thinking to leave them out as we have total functions, but maybe its a good idea to keep that in.

Reply:
I think these rules make the rules on line 36 and 40 redundant, so I would remove the rules on lines 36 and 40, and move these to those blocks of rules.